src/sonic-dhcp-server/dhcp_server/dhcprelayd.py

```diff
@@ -0,0 +1,182 @@
+        if op == "SET":
+            for field, value in entry:
+                if field != "state":
+                    continue
+                if key in self.dhcp_interfaces_state and value == self.dhcp_interfaces_state[key]:
+                    return
+                break
+        self.refresh_dhcrelay()
+
+    def _vlan_update_event(self):
+        key, _, _ = self.subscribe_vlan_table.pop()
+        # For vlan doesn't have related dhcp entry, not need to refresh dhcrelay process
+        if key not in self.dhcp_interfaces_state:
+            return
+        self.refresh_dhcrelay()
+
+    def _start_dhcrelay_process(self, new_dhcp_interfaces, dhcp_server_ip):
+        # To check whether need to kill dhcrelay process
+        kill_res = self._kill_exist_relay_releated_process(new_dhcp_interfaces, "dhcrelay")
+        if kill_res == NOT_KILLED:
+            # Means old running status consistent with the new situation, no need to run new
+            return
+
+        # No need to start new dhcrelay process
+        if len(new_dhcp_interfaces) == 0:
+            return
+
+        cmds = ["/usr/sbin/dhcrelay", "-d", "-m", "discard", "-a", "%h:%p", "%P", "--name-alias-map-file", "/tmp/port-name-alias-map.txt"]
+        for dhcp_interface in new_dhcp_interfaces:
+            cmds += ["-id", dhcp_interface]
+        cmds += ["-iu", "docker0", dhcp_server_ip]
+        subprocess.Popen(cmds)
+
+    def _start_dhcpmon_process(self, new_dhcp_interfaces):
+        # To check whether need to kill dhcrelay process
+        kill_res = self._kill_exist_relay_releated_process(new_dhcp_interfaces, "dhcpmon")
+        if kill_res == NOT_KILLED:
+            # Means old running status consistent with the new situation, no need to run new
+            return
+
+        # No need to start new dhcrelay process
+        if len(new_dhcp_interfaces) == 0:
+            return
+
+        for dhcp_interface in new_dhcp_interfaces:
+            cmds = ["/usr/sbin/dhcpmon", "-id", dhcp_interface, "-iu", "docker0", "-im", "eth0"]
+            subprocess.Popen(cmds)
+
+    def _kill_exist_relay_releated_process(self, new_dhcp_interfaces, process_name):
+        old_dhcp_interfaces = set()
+        target_proc = None
+
+        # Get old dhcrelay process and get old dhcp interfaces
+        for proc in psutil.process_iter():
+            if proc.name() == process_name:
+                cmds = proc.cmdline()
+                index = 0
+                target_proc = proc
+                while index < len(cmds):
+                    if cmds[index] == "-id":
+                        old_dhcp_interfaces.add(cmds[index + 1])
+                        index += 2
+                    else:
+                        index += 1
+        if target_proc is None:
+            return NOT_FOUND_PROC
+
+        # No need to kill
+        if process_name == "dhcrelay" and old_dhcp_interfaces == new_dhcp_interfaces or \
```

Comment:
Will it trigger ProcessMon alert?

Reply:
I don't think it would, because these two processes (dhcpmon and dhcrelay) have been removed from critical_processes